pkg/policy/k8s/service.go

```diff
@@ -0,0 +1,278 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package k8s
+
+import (
+	"errors"
+	"net/netip"
+	"slices"
+
+	"github.com/sirupsen/logrus"
+	"go4.org/netipx"
+	"k8s.io/apimachinery/pkg/labels"
+
+	ipcacheTypes "github.com/cilium/cilium/pkg/ipcache/types"
+	"github.com/cilium/cilium/pkg/k8s"
+	"github.com/cilium/cilium/pkg/k8s/types"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/policy/api"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+// isSelectableService returns true if the service svc can be selected by a ToServices rule
+// Normally, only services without a label selector (i.e. empty services)
+// are allowed as targets of a toServices rule.
+// This is to minimize the chances of a pod IP being selected by this rule, which might
+// cause conflicting entries in the ipcache.
+//
+// This requirement, however, is dropped for HighScale IPCache mode, because pod IPs are
+// normally excluded from the ipcache regardless. Therefore, in HighScale IPCache mode,
+// all services can be selected by ToServices.
+func (p *PolicyWatcher) isSelectableService(svc *k8s.Service) bool {
+	if svc == nil {
+		return false
+	}
+	return p.config.EnableHighScaleIPcache || svc.IsExternal()
+}
+
+// onServiceEvent processes a ServiceNotification and (if necessary)
+// recalculates all policies affected by this change.
+func (p *PolicyWatcher) onServiceEvent(event k8s.ServiceNotification) {
+	err := p.updateToServicesPolicies(event.ID, event.Service, event.OldService)
+	if err != nil {
+		p.log.WithError(err).WithFields(logrus.Fields{
+			logfields.Event:     event.Action,
+			logfields.ServiceID: event.ID,
+		}).Warning("Failed to recalculate CiliumNetworkPolicy rules after service event")
+	}
+}
+
+// updateToServicesPolicies is to be invoked when a service has changed (i.e. it was
+// added, removed, its endpoints have changed, or its labels have changed).
+// This function then checks if any of the known CNP/CCNPs are affected by this
+// change, and recomputes them by calling resolveCiliumNetworkPolicyRefs.
+func (p *PolicyWatcher) updateToServicesPolicies(svcID k8s.ServiceID, newSVC, oldSVC *k8s.Service) error {
+	var errs []error
+
+	// Bail out early if updated service is not selectable
+	if !(p.isSelectableService(newSVC) || p.isSelectableService(oldSVC)) {
+		return nil
+	}
+
+	// If the service selector or its labels have changed,
+	// we have to also want to update any CNPs that matched oldSVC
+	checkOldService := oldSVC != nil && !newSVC.DeepEqual(oldSVC)
+
+	for key, cnp := range p.cnpCache {
```

Comment:
In high-scale ipcache mode, we will be calling this function a **lot**. Specifically, every time a service changes backend IPs.

We should probably maintain a cache of CNPs / CCNPs that reference services. That way we can make this loop as short as possible.

Reply:
On second thought, a mapping of service -> CNP / CCNP names is probably the right thing here; we want to be able to cheaply handle updates here.

Reply:
Yeah, fair point. We will still have to iterate over all CNP/CCNP with a service label selector if a service label changes, but for everything else (which I expect to be the more common case) a `svc -> cnp`  map makes sense.

Reply:
Oh right, we have selectors. In that case, a list of policies with selectors would probably be a good easy optimization.